pkg/tools/etcd_helper.go

```diff
@@ -116,19 +127,70 @@ func (h *EtcdHelper) decodeNodeList(nodes []*etcd.Node, slicePtr interface{}) er
 			}
 			continue
 		}
-		obj := reflect.New(v.Type().Elem())
-		if err := h.Codec.DecodeInto([]byte(node.Value), obj.Interface().(runtime.Object)); err != nil {
-			return err
-		}
-		if h.Versioner != nil {
-			// being unable to set the version does not prevent the object from being extracted
-			_ = h.Versioner.UpdateObject(obj.Interface().(runtime.Object), node)
+		if obj, found := h.getFromCache(node.ModifiedIndex); found {
+			v.Set(reflect.Append(v, reflect.ValueOf(obj)))
+		} else {
+			obj := reflect.New(v.Type().Elem())
+			if err := h.Codec.DecodeInto([]byte(node.Value), obj.Interface().(runtime.Object)); err != nil {
+				return err
+			}
+			if h.Versioner != nil {
+				// being unable to set the version does not prevent the object from being extracted
+				_ = h.Versioner.UpdateObject(obj.Interface().(runtime.Object), node)
+			}
+			v.Set(reflect.Append(v, obj.Elem()))
+			if node.ModifiedIndex != 0 {
+				h.addToCache(node.ModifiedIndex, obj.Interface())
+			}
 		}
-		v.Set(reflect.Append(v, obj.Elem()))
 	}
 	return nil
 }
 
+// etcdCache defines interface used for caching objects stored in etcd. Objects are keyed by
+// their Node.ModifiedIndex, which is unique across all types.
+// All implementations must be thread-safe.
+type etcdCache interface {
+	getFromCache(index uint64) (interface{}, bool)
+	addToCache(index uint64, obj interface{})
+}
+
+const maxEtcdCacheEntries int = 100000
+
+func (h *EtcdHelper) getFromCache(index uint64) (interface{}, bool) {
+	var obj interface{}
```

Comment:
>Fortunately, the work being done for conversion can be used to implement efficient DeepCopy, so I would expect that portion to go down.

Yeah, but that work should also make this caching mechanism mostly unnecessary, after that is finished this will only save on unmarshalling costs.